src/python/pants/engine/README.md

```diff
@@ -49,44 +50,44 @@ then the `@rule` will run exactly once, and the value that it produces will be a
 
 #### Example
 
-As a very simple example, you might register the following `@rule` that can compute a `String`
-Product given a single `Int` argument.
+As a very simple example, you might register the following `@rule` that can compute a `str`
+`Product` given a single `int` argument.
 
 ```python
-@rule(str, [int])
-def int_to_str(an_int):
-  return str(an_int)
+@rule
+def int_to_str(value: int) -> str:
+  return str(value)
 ```
 
-The first argument to the `@rule` decorator is the `Product` (ie, return) type for the `@rule`. The
-second argument is a list of "parameter selectors" that declare the types of the input parameters for
-the `@rule`. In this case, because the `Product` type is `str` and there is one parameter
-selector (for `int`), this `@rule` represents a conversion from `int` to `str`, with no other inputs.
+Here the only requirements for the `@rule` are:
+1. The decorated function is a free, top-level function; ie: methods cannot be `@rule`s.
+2. The decorated function must be fully type annotated.
+3. As mentioned above, the types of the function parameters must be hashable and have useful equality.
 
-When the engine encounters this `@rule` while compiling the rule graph for `str`-producing-`@rules`,
+When the engine encounters this `@rule` while compiling the rule graph for `str`-producing-`@rule`s,
 it will next go hunting for the dependency `@rule` that can produce an `int` using the fewest number
 of `Params`. For example, if there was an `@rule` that could produce an `int` without consuming any
-`Params` at all (ie, a singleton), then that `@rule` would always be chosen first. If all `@rules` to
-produce `int`s required at least one `Param`, then the engine would next see whether the input `Params`
-contained an `int`, or whether there were any `@rules` that required only one `Param`, then two
-`Params`, and so on.
+`Params` at all (ie, a singleton), then that `@rule` would always be chosen first. If all `@rule`s
+to produce `int`s required at least one `Param`, then the engine would next see whether the input
+`Params` contained an `int`, or whether there were any `@rule`s that required only one `Param`, then
+two `Params`, and so on.
 
-In cases where this search detects any ambiguity (generally because there are two or more `@rules` that
-can provide the same product with the same number of parameters), rule graph compilation will fail with
-a useful error message.
+In cases where this search detects any ambiguity (generally because there are two or more `@rule`s
+that can provide the same product with the same number of parameters), rule graph compilation will
+fail with a useful error message.
```

Comment:
Here is where the colloquialism "rulesets" might come in handy, because the ambiguity can occur not just from having aliasing for specific `@rule` product/output pairs, but as a property of a whole subgraph. This distinction hasn't come up a ton yet in practice yet except in the native backend due to the (currently) small number of `@rule`s in the codebase. My first stab at trying to allude to that:
```suggestion
fail with a useful error message. Note that sometimes ambiguity can also occur as a result of sets of `@rule`s which produce two separate paths through the rule graph with overlapping parameters and output product, without any specific `@rule` or pair of `@rule`s causing the ambiguity.
```

Reply:
I'm going to defer on out of scope changes. I only:
+ Fixed 100 col line length limit.
+ Fixed @rule syntax guidance.

Reply:
That sounds great! Thank you!